feat: 토큰 생성 및 검증 기능을 구현한다

kiarakim opened the pull request:
## 📄 Summary
> OAuth에서 받아온 사용자를 저장한 상태에서
인증된 사용자의 id로 토큰 생성하는 기능과 토큰을 받았을 떄 그 토큰에서 id 추출해서 사용자 객체를 반환하는 기능을 구현했습니다.
우선 `api`는 임시로 작성한 것이고, `토큰 유효기간` 또한 논의해봐야 할 문제인 것 같습니다.

## 🕰️ Actual Time of Completion
>

## 🙋🏻 More
> 


close #88

feb-dain committed:
feat: 플렉스 박스 스타일 컴포넌트를 구현한다 (#173)

* test: 필요없는 파일 및 코드 삭제

[#114]

* test: 타입 중복 제거

- 재사용되는 타입은 전역으로 분리
- 컴포넌트의 Props export하여 스토리북에서 재사용

[#114]

* feat: FlexBox 컴포넌트 구현

[#114]

* design: GlobalStyle에서 `font` 적용하는 선택자 변경

이유
- font가 적용 안 되는 상황 방지
- 스토리북 font도 Noto Sans KR로 변경

[#114]

* chore: storybook 접근성 검사 addon 설치

[#114]

* refactor: 공통으로 사용되는 스타일 전역으로 분리

[#114]

* fix: 공통 style 전역으로 분리하면서 생기는 에러 개선

절대 경로 설정

[#114]

* refactor: 타입 이름 명확하게 변경

[#114]

* refactor: `justify content`와 `align items` 속성 따로 적용할 수 있도록 변경

[#114]

* test: row gap과 column gap 설명 보강

[#114]

* refactor: px → rem 단위로 변경

[#114]

kiarakim committed:
chore: spring security 의존성 추가

kiarakim committed:
chore: Json Web Token 의존성 추가

feb-dain committed:
refactor: FlexBox에 width, height 속성을 추가한다 (#178)

* refactor: FlexBox에 width, height 속성 추가

[#177]

* refactor: width/height를 적용하지 않았을 때, 어떤 값이 적용되는지 코드만 봐도 알 수 있도록 개선

[#177]

kiarakim committed:
feat: Token 생성 기능 구현

kiarakim committed:
feat: Token 검증 기능 구현

feb-dain committed:
fix: 스토리북 svg 경로 에러를 해결한다 (#184)

* refactor: px → rem으로 변경

[#175]

* fix: `@assets` 절대 경로 스토리북에 추가

[#183]

* chore: @svgr/webpack 설치

[#183]

* fix: storybook svg 경로 에러 개선

[#183]

kyw0716 committed:
fix: @style 경로 참조 에러 관련 웹팩 설정 추가 (#190)

[#189]

kyw0716 committed:
feat: 충전소 정보 서버 필터링 api를 mocking한다 (#188)

* feat: 서버 필터링 api mocking

[#169]

* refactor: 서버 필터링 부분 코드 가독성 개선

[#169]

* refactor: ,를 통해 여러 query param을 전달할 수 있도록 하는 유틸함수 추가

- URLSearchParams는 ,를 사용할 수 없어 새로운 유틸함수 추가함

[#169]

* feat: 간단한 예시 코드 추가

[#169]

* refactor: 실제 반환 데이터를 기반으로 mock data 생성하도록 수정

[#169]

* refactor: 충전기 타입, 회사명 상수화

[#169]

* feat: 충전소 세부정보에 상수화 된 충전기 타입 적용

[#169]

* refactor: 불필요한 console.log 삭제

[#169]

* refactor: 미리 상수화 해둔 객체를 활용해 mock data를 생성하도록 수정

[#169]

* refactor: url 변환 함수 코드 개선

[#169]

* fix: 잘못된 상수 수정

[#169]

kyw0716 committed:
feat: 네비게이터 컴포넌트를 구현한다 (#191)

* feat: Navigator 컴포넌트 추가

[#185]

* design: StationList 컴포넌트 위치 수정

[#185]

* feat: 지도에 Navigator 컴포넌트 추가

[#185]

* feat: Navigator에 아이콘 추가

[#185]

* design: cursor pointer 속성 추가

[#185]

* refactor: 로고 변경

[#185]

* refactor: px → rem

[#185]

kyw0716 committed:
feat: 실시간 필터링 옵션에 외부인 출입 제한 여부에 대한 선택 옵션 추가 (#193)

[#186]

gabrielyoon7 committed:
refactor: Text variant type 지정 (#194)

[#192]

drunkenhw committed:
feat: 충전소를 고장 신고하는 기능을 구현한다 (#196)

* feat: 고장 신고 클래스 추가

[#180]

* feat: 고장 신고 저장하는 기능 추가

[#180]

* feat: 고장 신고 기능 추가

[#180]

* feat: 고장 신고 api 추가

[#180]

* feat: 이미 신고한 충전소 검증 기능 추가

[#180]

gabrielyoon7 committed:
feat: List와 ListItem 컴포넌트를 구현한다 (#197)

* feat: List 컴포넌트 추가

[#172]

* feat: ListItem 컴포넌트 추가

[#172]

* fix: 오류로 인해 사라진 ListItem 컴포넌트 복구

[#172]

* feat: List Storybook 및 ListItem readme 추가

* feat: List에 디자인 옵션 추가

[#172]

* feat: ListItem에 clickable과 divider 추가

[#172]

* test: Menu 예제 추가

[#172]

* fix: Text 스토리타입 타입 오류 수정

[#172]

* test: List로 제어할 수 있는 예제 추가

[#172]

* feat: CSSProp 기능 추가

[#172]

* test: ListItem 스토리북 예제 구현

[#172]

* refactor: Text mb가 0일 때 아무것도 하지 않도록 개선

[#172]

gabrielyoon7 committed:
feat: 사용자 식별이 가능하도록 토큰을 생성하는 기능 구현 (#199)

* feat: 최초 접속 시 랜덤한 8자리 정수의 토큰을 발행하여 로컬 스토리지에 저장하는 기능 구현

- 단, 이미 발행된 토큰이 발견되는 경우에는 재발급 하지 않음

* refactor: 불필요한 콘솔 제거

* refactor: 스토리지 키 위치 수정

[#198]

feb-dain committed:
refactor: 절대 경로 확장성 있게 수정한다 (#205)

[#203]

feb-dain committed:
chore: `@heroicons/react`를 설치한다 (#202)

[#200]

feb-dain committed:
chore: `@heroicons/react`를 설치한다 (#202)

[#200]

kiarakim committed:
chore: oAuth 의존성 추가

kiarakim committed:
feat: Member 엔티티 및 테스트 코드 작성

kiarakim committed:
feat: 인증 서비스 작성

kiarakim committed:
feat: id로 토큰 생성 기능 구현

kiarakim committed:
test: 토큰 만료 검증 테스트 작성

kiarakim committed:
comment: 주석 및 로그 제거

kiarakim committed:
feat: 토큰의 id로 DB Member 반환 기능 구현

kiarakim committed:
feat: Controller 작성

gabrielyoon7 committed:
feat: develop mode에서 msw와 로컬 서버를 동적으로 선택하는 기능을 구현한다 (#215)

* refactor: msw를 실행하는 함수 분리

[#210]

* feat: msw를 키고 끄는 버튼 추가

[#210]

* feat: msw 컨트롤 버튼이 development 모드에서만 보이도록 수정

[#210]

* docs: 프론트엔드 설치 및 실행 방법 설명 추가

[#210]

* refactor: isMswMode로 상태명 변경

[#210]

drunkenhw commented on backend/build.gradle:
```diff
@@ -31,6 +31,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
     implementation 'org.springframework.boot:spring-boot-starter-validation'
     implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
spring security 라이브러리를 사용하면 안되는 것으로 알고 있는데 혹시 이 라이브러리를 빼고 구현할 수는 없을까요?

kiarakim replied:
뺐어요

drunkenhw commented on backend/build.gradle:
```diff
@@ -31,6 +31,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-security'
+    implementation 'io.jsonwebtoken:jjwt:0.9.1'
+    implementation 'org.springframework.boot:spring-boot-starter-oauth2-client'
```
이 라이브러리도 빼는 것이 좋을 것 같습니다

kiarakim replied:
뺏어요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/config/AuthenticationConfig.java:
```diff
@@ -0,0 +1,53 @@
+@Configuration
+@EnableWebSecurity
+@RequiredArgsConstructor
+public class AuthenticationConfig implements WebMvcConfigurer {
```
해당 클래스도 spring security의 클래스 같은데 제거되어야 할 것 같아요

kiarakim replied:
제거했어여

drunkenhw replied:
Authentication Config라는 클래스 이름보다 그냥 WebConfig 이용해서 좀 더 유연한 클래스처럼 보이는 것도 좋을 것 같아요!

kiarakim replied:
녜에

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/controller/security/SearchController.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/search")
+public class SearchController {
+    @PostMapping
+    public ResponseEntity<MemberResponse> startSearching(@AuthMember Member member) {
+        MemberResponse response = new MemberResponse(member.getId(), member.getEmail());
+        return ResponseEntity.ok().body(response);
+    }
+}
```
무슨 역할을 하는 컨트롤러인지 잘 모르겠습니다..

kiarakim replied:
사라진 클래스입니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/controller/security/UserController.java:
```diff
@@ -0,0 +1,24 @@
+
+    @PostMapping("/jwt")
+    public ResponseEntity<String> createJwt() {
+//        String jwt = authService.createJwt(loginRequest.userName()); // DTO 로 response 만들어서 반환
+//        return ResponseEntity.ok().body(jwt);
```
주석은 제거 부탁드리겠습니다

kiarakim replied:
넵

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/AuthMemberResolver.java:
```diff
@@ -0,0 +1,51 @@
+        HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();
+        String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
+
+        String token = authorization.substring(TOKEN_START_INDEX);
```
Token start index는 맞지만 만약 `Bearer ~!~~!~!~` 이런식이 아닌 
`Brrrrr ~!2121~`이런식의 토큰도 호환가능할 것 같은데 좀 더 사용자 인증과 관련된 부분이니 좀 더 빡빡하게 하는 것이 좋지 않을까요?

kiarakim replied:
startsWith으로 확인해주었습니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/AuthMemberResolver.java:
```diff
@@ -0,0 +1,51 @@
+        String token = authorization.substring(TOKEN_START_INDEX);
+        Long memberId = JwtUtil.extractId(token, secretKey);
+
+        return memberRepository.findById(memberId);
```
Member 도메인 객체를 반환하는 이유가 궁금합니다 순전히 궁금증입니다 알려주세요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/JwtFilter.java:
```diff
@@ -0,0 +1,54 @@
+@Slf4j
+public class JwtFilter extends OncePerRequestFilter {
+
+    private static final int TOKEN_START_INDEX = 7;
+    private static final String BEARER_PREFIX = "Bearer ";
+    private static final String ROLE_USER = "ROLE_USER";
+    private final AuthService authService;
+    private final String secretKey;
```
상수와 필드 사이의 개행 부탁드려요

kiarakim replied:
넵

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/JwtFilter.java:
```diff
@@ -0,0 +1,54 @@
+
+        Long id = JwtUtil.extractId(token, secretKey);
+
+        UsernamePasswordAuthenticationToken authenticationToken =
+                new UsernamePasswordAuthenticationToken(id, null, List.of(new SimpleGrantedAuthority(ROLE_USER)));
+
+        authenticationToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
+        SecurityContextHolder.getContext().setAuthentication(authenticationToken);
```
해당 부분은 어떻게 동작하는 것이죠?

kiarakim replied:
잘 몰라서 제거했습니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/JwtUtil.java:
```diff
@@ -0,0 +1,39 @@
+
+        return Jwts.builder()
+                .setClaims(claims)
+                .setIssuedAt(new Date(System.currentTimeMillis()))
+                .setExpiration(new Date(System.currentTimeMillis() + expiredMs))
```
Date를 LocalDateTime으로 사용하시지 않은 이유가 있나요?
https://docs.oracle.com/javase/8/docs/api/java/util/Date.html
Date는 Deprecate 됐다고 하네요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/JwtUtil.java:
```diff
@@ -0,0 +1,39 @@
+
+import java.util.Date;
+
+public class JwtUtil {
```
Util이라는 네이밍보다 좀 더 괜찮은 것은 없을까요?
그리고 해당 클래스를 static 메서드를 가진 객체가 아닌 `@Component`로 등록하지 않으신 이유가 있을까요?

kiarakim replied:
Bean을 잘 몰랐어서.. 이 클래스는 사라졌습니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/service/security/AuthService.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final MemberRepository memberRepository;
+
+    @Value("${jwt.secret}")
+    private String secretKey;
```
secret key가 JwtUtil이라는 클래스에 있는 것이 좀 더 유지보수에 좋을 것 같아요

kiarakim replied:
네 고쳤습니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/service/security/AuthService.java:
```diff
@@ -0,0 +1,29 @@
+    @Value("${jwt.secret}")
+    private String secretKey;
+
+    private Long expiredMs = 1000 * 60 * 60L;
```
이 부분도요

kiarakim replied:
yml 파일에 암호화해놨습니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/service/security/AuthService.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Transactional
+    public String makeJwt() {
+        Member member = new Member("aaa@gmail.com");
+        memberRepository.save(member);
+        Long id = member.getId();
+        return JwtUtil.createJwtById(id, secretKey, expiredMs);
```
갑자기 aaa라는 회원을 저장하는 이유가 뭔지 궁금해요

kiarakim replied:
임시 데이터였는데 push할 때 잊었던 것같ㅇ아요 . 주의하겠ㅅ브니ㅣ다

drunkenhw commented on backend/src/main/resources/application.yml:
```diff
@@ -16,3 +16,6 @@ api:
 initialize-charge:
   enabled: true
+
+jwt:
+  secret: ${JWT_KEY}
```
이 key도 환경변수가 아닌 jasypt로 관리하는 것은 어떤가요

kiarakim replied:
그럼 key를 key로 암호화하는건데 고건 생각 못했네요. 괜찮은 것같아요.

drunkenhw commented on backend/src/test/java/com/carffeine/carffeine/station/domain/security/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+class MemberRepositoryTest {
```
중요한 어노테이션이 클래스이름이랑 더 가깝게 왔으면 좋겠어요

kiarakim replied:
네에

drunkenhw commented on backend/src/test/java/com/carffeine/carffeine/station/domain/security/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Autowired
+    private MemberRepository memberRepository;
+
+    private String email;
+    private Long memberId;
+
+    @BeforeEach
+    void setUp() {
+        email = "xxx@gmail.com";
+        Member member = new Member(email);
+        Member save = memberRepository.save(member);
+        memberId = save.getId();
+    }
```
setup으로 해당 member를 저장하는 이유가 있나요?

drunkenhw commented on backend/src/test/java/com/carffeine/carffeine/station/infrastructure/api/JwtUtilTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    void 유효기간이_1초인_토큰은_1초_뒤_만료된다() throws InterruptedException {
+        // given
+        expiredMs = 1000L;
+        String token = JwtUtil.createJwtById(id, secretKey, expiredMs);
+
+        // when
+        Thread.sleep(3000L);
+
+        // then
+        assertThatThrownBy(() -> JwtUtil.isExpired(token, secretKey));
+    }
```
테스트가 3초간 쉬는 것은 비효율적인 것 같아요 시간을 더 짧게 가져가는 방법은 어떠신가요?

drunkenhw reviewed (changes requested):
열심히 리뷰 달았습니다.
하지만 저희 우테코에서 spring security와 같은 기술은 아마 기술 검토에서 허락을 맡아야 사용가능한 것으로 알고있어서 해당 부분은 아마 수정해야할 것 같습니다.
수고하셨습니다!~

kyw0716 committed:
feat: 서버 필터링 옵션을 선택하는 UI를 구현한다 (#207)

* refactor: FilterButtonList → StationFilterButtonList로 네이밍 변경

[#187]

* feat: 서버 필터링 기능 선택 UI 추가

[#187]

* feat: Navigator 컴포넌트에서 ServerStationFilters 컴포넌트 열고 닫는 기능 구현

[#187]

* refactor: StationFilterButtonList → ClientStationFilters로 네이밍 변경

[#187]

* refactor: 충전 속도 타입 추가

[#187]

* feat: 지도에 ServerStationFilter 컴포넌트 추가

[#187]

* design: 필터링 컴포넌트 우측 border 추가

[#187]

* feat: 서버 필터링 선택 UI 추가

[#187]

* refactor: FlexBox 컴포넌트 props 추가 → alignItems

[#187]

* fix: 잘못 들어간 타입 수정

[#187]

* style: props 넘길 때 '' → {''}로 수정

[#187]

* feat: 서버 필터링 옵션 적용하기 기능 구현

[#187]

* refactor: map을 참조하는 방식 수정

- 기존에 map을 인자로 넘겨 받아 수행하던 작업들 store에서 직접 참조하는 방식으로 변경

[#187]

* refactor: 필터링 버튼 사이의 간격 추가

[#187]

* design: 아이콘 크기 수정

[#187]

* refactor: 색상 표기 방식 변경 black → #000

[#187]

* refactor: 절대 경로 수정

[#187]

* design: 우아한 다크모드 삭제

[#187]

gabrielyoon7 committed:
feat: Box 컴포넌트를 구현한다 (#219)

* feat: Box 컴포넌트 추가

[#217]

* feat: border 추가

[#217]

* test: border 예제 추가

[#217]

* feat: p 기능 구현

[#217]

* feat: p 기능 완성

[#217]

* feat: m 기능 완성

[#217]

* feat: 높이, 너비 기능 완성

[#217]

* test: 높이, 너비 테스트

[#217]

* feat: bgColor, color 구현

[#217]

* test: bgColor, color 스토리북 작성

[#217]

* feat: position 관련 기능 추가

[#217]

* test: position 관련 기능 예제 추가

[#217]

* test: 부족한 description 추가

[#217]

* test: 부족한 description 추가

[#217]

* refactor: 리뷰 반영

[#217]

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/station/controller/security/SearchController.java:
```diff
@@ -0,0 +1,19 @@
+@RestController
+@RequestMapping("/api/search")
+public class SearchController {
+    @PostMapping
```
개행 추가해주세요~

kiarakim replied:
넹. 근데 사라진 클래스..ㅠㅠ

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/station/controller/security/UserController.java:
```diff
@@ -0,0 +1,24 @@
+    public ResponseEntity<String> createJwt() {
+//        String jwt = authService.createJwt(loginRequest.userName()); // DTO 로 response 만들어서 반환
+//        return ResponseEntity.ok().body(jwt);
+        String jwt = authService.makeJwt();
```
사용자의 정보는 없어도 될까요?

kiarakim replied:
사용자 정보는 로그인서비스에서 member 객체 만드는데 거기서 바로 jwt 만들면 돼서요..!

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/station/domain/security/MemberRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface MemberRepository extends Repository<Member, Long> {
+
+    Member findByEmail(String email);
```
반환형이 `Optional<Member>`가 되면 어떨까요? 혹시 `Member`로 반환형을 주는 이유가 있을까요?

kiarakim replied:
이미 로그인 되어 DB에 저장한 사람의 토큰을 검증할 때만 사용해서 Optional를 고려하지 않았는데 제가 생각못한 반례가 있을까욥

drunkenhw committed:
feat: 고장 신고 여부를 확인하는 기능을 구현한다  (#201)

* feat: 고장 신고를 회원과 충전소 기준으로 찾는 기능 추가

[#181]

* feat: 고장 신고를 회원과 충전소 기준으로 찾는 API 추가

[#181]

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/JwtFilter.java:
```diff
@@ -0,0 +1,54 @@
+            return;
+        }
+
+        Long id = JwtUtil.extractId(token, secretKey);
```
JwtUtil을 빈으로 등록해서 사용하는 것은 어떨까요?

kiarakim replied:
해당 클래스는 삭제됐고 위 기능을 Jwt 클래스가 하고있어욥

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/config/AuthenticationConfig.java:
```diff
@@ -0,0 +1,53 @@
+    private String secretKey;
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity httpSecurity) throws Exception {
```
ㅠㅠ 스프링 시큐리티네요.. 나중에 제거하시고 필터로 직접 만들어주세요~

kiarakim replied:
네ㅠㅠ 반여완요

sosow0212 commented on backend/src/main/java/com/carffeine/carffeine/config/AuthenticationConfig.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.util.List;
+
+@Configuration
```
@Configuration의 위치가 클래스 바로 위로 가는 건 어떨까요?
중요한 어노테이션일수록 클래스 이름과 가깝게요

kiarakim replied:
넵 감사합닏

sosow0212 reviewed (changes requested):
작업하시느라 수고하셨습니다👍
스프링 시큐리티를 사용하셔서.. 다시 덜어내고 작업 하셔야할 것 같아요 ㅠㅠ
남긴 리뷰 몇 가지 확인해주세요~

++ 추가적으로 CI에서 테스트 실패로 뜨네요 확인 부탁드릴게요!

kiarakim committed:
refactor: 토큰 생성 및 검증 기능 수정

drunkenhw committed:
feat: 스토리북 배포를 자동화한다 (#206)

kiarakim committed:
refactor: User 도메인에 id 추가

gabrielyoon7 committed:
feat: TextField를 구현한다 (#222)

* fix: 잘못된 속성 제거

[#220]

* feat: 기본 Input 컴포넌트 작성

[#220]

* feat: TextField로 변경 및 디자인 추가

[#220]

* refactor: 불필요한 컴포넌트 제거

[#220]

* feat: placeHolder 기능 추가 및 css 프롭 분리

[#220]

* feat: value, onChange 추가

[#220]

* feat: helperText 구현

[#220]

* test: description 개선

[#220]

* feat: width 기능 구현

[#220]

* refactor: 피드백 반영

[#220]

* refactor: 함수 순서 변경

[#220]

* refactor: 피드백 반영

[#220]

feb-dain committed:
feat: 검색창 컴포넌트를 구현한다 (#221)

* refactor: px → rem으로 변경

[#175]

* feat: StationSearch 컴포넌트 구현

[#175]

* refactor: getSize 함수 전역으로 관리

[#175]

* refactor: Button에 width, height 속성 추가

[#175]

* refactor: ListItem에 필요없는 속성 제거하고 NoLastDivider 속성 추가

[#175]

* refactor: gitSize 전역으로 관리

[#175]

* refactor: Text 타입 깔끔하게 분리

[#175]

* refactor: 검색 컴포넌트 구현

[#175]

* refactor: 검색바 분리 및 속성 추가

[#175]

* refactor: Text 스타일 컴포넌트에 새로운 속성 추가

[#175]

* refactor: 스타일 아래로 이동

[#175]

* refactor: 필요없는 코드 제거

[#175]

* feat: 검색 결과 리스트 컴포넌트 구현

[#175]

* feat: 충전소 검색창 컴포넌트 구현

[#175]

kyw0716 committed:
feat: 누적 혼잡도 통계를 시각화 해주는 컴포넌트를 개발한다 (#223)

* refactor: 불필요한 import 삭제

* refactor: 허스키 삭제

* refactor: 필터 적용하기 버튼 sticky 적용

* feat: 수치에 따라 길이가 변하는 그래프 바 구현

* feat: 혼잡도 통계 그래프 UI 구현

- 이전에 구현한 그래프 바 이용
- 헤더에 요일 버튼 UI 구현
- 그래프 바에 시간 정보 추가

* feat: 누적 혼잡도 ratio가 -1일 경우 회색으로 표시하는 기능 구현

* refactor: 구조분해 할당 적용

* feat: 급속/완속, 요일 변경 기능 추가

* feat: 선택된 요일 표시 기능 추가

* refactor: 급속/완속 토글 메서드 코드 개선

- if 분기처리 → 삼항연산자

kiarakim committed:
feat: Member의 role 제거

kiarakim committed:
refactor: 토큰 생성 방식 변경

kiarakim committed:
refactor: 토큰 검증 방식 변경

kiarakim committed:
refactor: Jwt 객체 사용

kiarakim committed:
test: 빈 상속 클래스 생성

feb-dain committed:
feat: pr시 comment에 스토리북 배포 링크를 자동으로 달아주는 기능을 구현한다 (#227)

[#226]

kiarakim committed:
chore: oauth 의존성 라이브러리 제거

kiarakim committed:
refactor: 사용하지 않는 파라미터 제거

kiarakim committed:
refactor: 패키지 명 변경

kiarakim committed:
refactor: 클래스 네이밍 변경 User -> Member

kiarakim committed:
refactor: 컨벤션 적용

kiarakim committed:
refactor: 메서드 분리

kiarakim committed:
refactor: Member객체 대신 id를 반환하도록 변경

kiarakim committed:
refactor: 어노테이션 선언 위치 변경

kiarakim committed:
refactor: 빈줄 컨벤션 적용

kiarakim committed:
test: mock bean 주입 클래스 상속

kiarakim committed:
Merge branch 'develop' into feat/88

be-student commented on backend/src/main/java/com/carffeine/carffeine/station/controller/auth/MemberController.java:
```diff
@@ -0,0 +1,21 @@
+    private final MemberService memberService;
+
+    @PostMapping("/jwt")
+    public ResponseEntity<String> createJwt() {
```
임시 api 인 것 같아요!
제거하면 좋지 않을까요?

kiarakim replied:
넴. 나중에 머지되면 사라진 클래스지만 제거하겟습니다

be-student commented on backend/src/main/java/com/carffeine/carffeine/station/controller/auth/SearchController.java:
```diff
@@ -0,0 +1,19 @@
+@RestController
+@RequestMapping("/api/search")
+public class SearchController {
+    @PostMapping
```
- 한줄을 띄면 좋을 것 같아요!

kiarakim replied:
이 클래스는 바람속으로 사라졌습니다. 감사합니다

be-student commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private void sendUnauthorizedError(HttpServletResponse response, String errorMessage) throws IOException {
+        response.sendError(HttpServletResponse.SC_UNAUTHORIZED, errorMessage);
```
여기에 직접 body 를 넣어줄 수 있는데요
message 만 전달하는 것이 아니라
백엔드에서 사용하는 errorResponse 였나 하는 그 클래스를 사용하면 어떨까요?

drunkenhw replied:
동감합니다

kiarakim replied:
반영햇습지다

be-student commented on backend/src/main/java/com/carffeine/carffeine/station/domain/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,54 @@
+
+    private Date getExpiration() {
+        LocalDateTime now = LocalDateTime.now();
+        return Date.from(now.plusHours(1).atZone(ZoneId.systemDefault()).toInstant());
```
추후에 1시간이라는 부분은 yml 로 관리하게 된다면 더 유연하게 사용할 수 있을 것 같아요

kiarakim replied:
햇어요

be-student reviewed (approved):
아주 고생 많으셨어요 👍 
처음 하는 작업이라서 정말 힘들었을 텐데 잘 해주셨네요

kiarakim committed:
refactor: 사용하지 않는 api 및 Controller 제거

sosow0212 reviewed (approved):
처음하는 작업이라 힘드셨을텐데 수고 많으셨습니다. 👍

kiarakim committed:
Merge branch 'develop' into feat/88

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,59 @@
+            return;
+        }
+
+        String token = authorization.substring(TOKEN_START_INDEX);
```
```suggestion
        String token = authorization.substring(BEARER_PREFIX.length());
```
이 방식은 어떠신가요?

kiarakim replied:
좋아요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,59 @@
+        String token = authorization.substring(TOKEN_START_INDEX);
+        Jwt jwt = new Jwt(token);
+
+        if (jwt.isExpired(token, secretKey)) {
+            filterChain.doFilter(request, response);
+        }
+
+        Long id = jwt.extractId(token, secretKey);
+
+        Member member = memberRepository.findById(id);
+        if (member == null) {
+            filterChain.doFilter(request, response);
+        }
```
이 부분을 if로 validation 할 때 `sendUnauthorizedError` 을 호출하는 것은 어떤가요?

kiarakim replied:
좋네요. 이런건 어떻게 알고 리뷰해주신거죠👍

kiarakim replied:
넴 반영완ㅇ료

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,59 @@
+    private static final String BEARER_PREFIX = "Bearer ";
+
+    private final MemberRepository memberRepository;
+    private final String secretKey;
```
secret key 를 한 곳에서 관리하는 것이 어떤가요?

kiarakim replied:
한곳에서 관리하도록 했습니다.

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/jwt/JwtProvider.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Date;
+
+@Component
+public class JwtProvider {
```
jwt provider와 jwt 클래스가 나뉘어져있는 이유가 뭔가요?

kiarakim replied:
그러게요. 붙여주었습니다.

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/exception/AuthorizationException.java:
```diff
@@ -0,0 +1,10 @@
+import com.carffeine.carffeine.common.exception.BaseException;
+import com.carffeine.carffeine.common.exception.ExceptionType;
+
+public class AuthorizationException extends BaseException {
+    public AuthorizationException(ExceptionType exceptionType) {
+        super(exceptionType);
+    }
```
개행 부탁드려요

kiarakim replied:
네 했어요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/exception/StationExceptionType.java:
```diff
@@ -7,7 +7,8 @@ public enum StationExceptionType implements ExceptionType {
     INVALID_LONGITUDE(Status.INVALID, 1002, "유효하지 않는 경도입니다"),
-    NOT_FOUND_ID(Status.NOT_FOUND, 1003, "요청하신 id의 충전소를 찾을 수 없습니다.");
+    NOT_FOUND_ID(Status.NOT_FOUND, 1003, "요청하신 id의 충전소를 찾을 수 없습니다."),
+    NO_AUTHORIZATION(Status.INVALID, 1004, "토큰의 권한이 없습니다.");
```
station Exception에  권한이 없다는 것은 이상한 것 같아요

kiarakim replied:
없앴어요

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/infrastructure/api/AuthMemberResolver.java:
```diff
@@ -0,0 +1,50 @@
+
+    private final MemberRepository memberRepository;
+
+    @Value("${jwt.secret}")
+    private String secretKey;
```
secret key가 한 군데에서 관리하면 좋을 것 같아요

kiarakim replied:
한곳에서 관리하도록 했습니다.

drunkenhw reviewed (changes requested):
리뷰 남겼습니다 감사합니다~

kiarakim committed:
refactor: 클래스 네이밍 변경

kiarakim committed:
refactor: secret을 한 곳에서 관리하도록 수정

kiarakim committed:
refactor: 개행 추가

kiarakim committed:
refactor: 사용하지 않는 인스턴스 제거

kiarakim committed:
refactor: JwtProvider의 기능을 Jwt에 통합

kiarakim committed:
refactor: 부분 문자열 추출 범위 수정

kiarakim committed:
refactor: 토큰 만료 기간 암호화

kiarakim committed:
refactor: 토큰 만료 기간 암호화

kiarakim committed:
refactor: Optional 처리 및 커스텀 예외 사용

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        String token = authorization.substring(BEARER_PREFIX.length());
+        Jwt jwt = new Jwt();
```
Jwt 클래스는 component인데 굳이 new 로 생성해주지말고 빈으로 주입하는 것이 좋아보여용

kiarakim replied:
녜에

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private boolean isMemberPresent(Long id) {
+        return memberRepository.findById(id)
```
member가 존재한다는 부분은 exists라는 메서드로 하면 좋을 것 같아요

kiarakim replied:
넹

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,67 @@
+        Jwt jwt = new Jwt();
+
+        if (jwt.isExpired(token)) {
+            filterChain.doFilter(request, response);
```
여기도 sendError~~~ 해주면 좋을 것 같아요

kiarakim replied:
했습니다 감사합니다

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/filter/JwtFilter.java:
```diff
@@ -0,0 +1,67 @@
+
+        Long id = jwt.extractId(token);
+
+        if (isMemberPresent(id)) {
+            sendUnauthorizedError(response, "등록되지 않은 회원입니다");
+            filterChain.doFilter(request, response);
```
멤버가 존재한다면 에러? 로직이 반대로 된 거 같아요

kiarakim replied:
그러게요 이건 제가 왜 그랬을까요

kiarakim committed:
refactor: 컴포넌트 클래스 의존성 주입으로 변경

drunkenhw commented on backend/src/main/java/com/carffeine/carffeine/station/domain/jwt/Jwt.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private String createToken(Claims claims) {
+        Key key = gethmacShaKey();
```
Key는 한번만 만들 수 없나요?

kiarakim replied:
이거 한번만 만든건디 이거 메서드에여

kiarakim replied:
아니네요. 고쳣습ㄴ디ㅏ

drunkenhw reviewed (changes requested):
열심히 리뷰했습니다 감샇반디ㅏ

kiarakim committed:
refactor: Key 한번만 생성하도록 변경

kiarakim committed:
refactor: 에러처리

drunkenhw reviewed (approved):
키아라수고하셨습니다고생하셧습니다리뷰많이해서죄송합니다
